### PR TITLE
Maintenance Part 3: Linode Detail Banner

### PR DESCRIPTION
**Pull request by martinmckenna**
## Description

Adds Banner to Linode Detail Screen. **There are a lot of different possible messaging cases, so please use all the mock data provided with prod-test-003**

## Type of Change
- Non breaking change ('update', 'change')

## Applicable E2E Tests

N/A

## Note

Mock Data

```js
{
  "results": 5,
  "page": 1,
  "data": [
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/500",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal1-vagrant"
      },
      "when": null,
      "body": null,
      "severity": "major",
      "type": "migration_scheduled",
      "label": "You have a scheduled migration pending!",
      "message": "You have a scheduled migration pending!  Your Linode will be automatically shut down, migrated, and returned to its last state."
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/504",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal5-xengrant"
      },
      "when": null,
      "body": null,
      "severity": "major",
      "type": "migration_pending",
      "label": "You have a migration pending!",
      "message": "You have a migration pending! Your Linode must be offline before starting the migration."
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/500",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal1-vagrant"
      },
      "when": "2019-06-29T15:32:26",
      "body": "This Linode resides on a host that is pending critical maintenance. You should have recieved a support ticket that details how you will be affected. Please see the aformentioned ticket and [status.linode.com](https://status.linode.com/) for more details.",
      "severity": "critical",
      "type": "maintenance",
      "label": "zombieload-migration-scheduled",
      "message": "This Linode will be affected by critical maintenance!"
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/500",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal1-vagrant"
      },
      "when": "2019-10-29T15:32:26",
      "body": "This Linode resides on a host that is pending critical maintenance. You should have recieved a support ticket that details how you will be affected. Please see the aformentioned ticket and [status.linode.com](https://status.linode.com/) for more details.",
      "severity": "critical",
      "type": "maintenance",
      "label": "zombieload-migration-scheduled",
      "message": "This Linode will be affected by critical maintenance!"
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/500",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal1-vagrant"
      },
      "when": "2019-02-28T15:32:26",
      "body": "This Linode resides on a host that is pending critical maintenance. You should have recieved a support ticket that details how you will be affected. Please see the aformentioned ticket and [status.linode.com](https://status.linode.com/) for more details.",
      "severity": "critical",
      "type": "maintenance",
      "label": "zombieload-migration-scheduled",
      "message": "This Linode will be affected by critical maintenance!"
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/507",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal8-xengrant"
      },
      "when": "2020-03-22T18:58:41",
      "body": "This Linode resides on a host that is pending critical maintenance. You should have recieved a support ticket that details how you will be affected. Please see the aformentioned ticket and [status.linode.com](https://status.linode.com/) for more details.",
      "severity": "critical",
      "type": "maintenance",
      "label": "zombieload-reboot-scheduled",
      "message": "This Linode will be affected by critical maintenance!"
    },
    {
      "until": null,
      "entity": {
        "url": "/linode/instances/507",
        "id": 14341983,
        "type": "linode",
        "label": "loadbal8-xengrant"
      },
      "when": "2019-03-22T18:58:41",
      "body": "This Linode resides on a host that is pending critical maintenance. You should have recieved a support ticket that details how you will be affected. Please see the aformentioned ticket and [status.linode.com](https://status.linode.com/) for more details.",
      "severity": "critical",
      "type": "maintenance",
      "label": "zombieload-reboot-scheduled",
      "message": "This Linode will be affected by critical maintenance!"
    }
  ],
  "pages": 1
}
```

### CODE REVIEW

File: src/features/linodes/LinodesDetail/LinodesDetailHeader/MigrationNotification.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  /** Migrate */
+  const migrate = () => {
+    scheduleOrQueueMigration(linodeID)
+      .then(_ => {
+        // A 200 response indicates that the operation was successful.
+        const successMessage =
+          notificationType === 'migration_scheduled'
+            ? 'Your Linode has been entered into the migration queue.'
+            : 'Your migration has been scheduled.';
```
Comment (martinmckenna) [inline src/features/linodes/LinodesDetail/LinodesDetailHeader/MigrationNotification.tsx]: nothing was changed in this file. just moved it into it's own component

File: src/components/MaintenanceBanner/MaintenanceBanner.tsx [hunk 2]
```diff
@@ -82,3 +95,70 @@ export default compose<CombinedProps, Props>(
...
+    } else {
+      /**
+       * for some reason, we don't have a _when_ property from the notification.
+       * In other words, we have no idea when their maintenance time starts. This can happen
+       * when the user has a manual migration pending - only a small amount of users are going
+       * to see this.
+       *
+       * @todo not sure what the messaging will be yet in this case.
+       */
+      return <React.Fragment>hello world</React.Fragment>;
```
Comment (martinmckenna) [inline src/components/MaintenanceBanner/MaintenanceBanner.tsx]: yeah this is probably going to be coming in the next PR sorry about this

File: src/components/MaintenanceBanner/MaintenanceBanner.tsx [hunk 3]
```diff
@@ -82,3 +95,70 @@ export default compose<CombinedProps, Props>(
   styled,
   React.memo
 )(MaintenanceBanner);
+
+const generateIntroText = (
+  type?: 'migration' | 'reboot',
+  start?: string | null,
```
Comment (Jskobos) [inline src/components/MaintenanceBanner/MaintenanceBanner.tsx]: are undefined and null separate cases that have to be handled for these?

Comment (martinmckenna) [reply src/components/MaintenanceBanner/MaintenanceBanner.tsx]: yes. If there is no `maintenance` notification, this Linode will not have the `maintenance` key.

BUT the API can also return `null` for the `when` key-value pair

File: src/components/MaintenanceBanner/MaintenanceBanner.tsx [hunk 4]
```diff
@@ -47,22 +62,33 @@ const MaintenanceBanner: React.FC<CombinedProps> = props => {
...
+   *
+   * This case will happen when there's a Linode with no MQueue record
+   * but has a pending manual migration.
+   *
+   * In other words, it's an edge case, and we don't need to worry about
+   * informing the user of anything in this case.
+   */
+  if (!maintenanceStart) {
+    return null;
+  }
```
Comment (martinmckenna) [inline src/components/MaintenanceBanner/MaintenanceBanner.tsx]: this code here is breaking the banners from appearing on the dashboard and linodes landing. fix coming in the next PR.

**Pull request merged**